pkg/scheduler/scheduler.go

```diff
@@ -390,79 +372,132 @@ func (s *Scheduler) scheduleNext() bool {
+		return nil
+	}
+	if features.FeatureGate.Enabled(features.Failover) {
+		klog.Infof("Reschedule ResourceBinding(%s/%s) as cluster failure", namespace, name)
+		err = s.rescheduleResourceBinding(rb)
+		metrics.BindingSchedule(string(FailoverSchedule), metrics.SinceInSeconds(start), err)
+		return err
+	}
+	return nil
+}
 
-		clusterPolicy, err := s.clusterPolicyLister.Get(clusterPolicyName)
-		if err != nil {
-			return err
+func (s *Scheduler) doScheduleClusterBinding(name string) error {
+	crb, err := s.clusterBindingLister.Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// the binding does not exist, do nothing
+			return nil
 		}
+		return err
+	}
 
-		return s.scheduleClusterResourceBinding(clusterResourceBinding, clusterPolicy)
+	start := time.Now()
+	if !helper.IsBindingReady(&crb.Status) {
+		// the binding has not been scheduled, need schedule
+		klog.Infof("Start scheduling ClusterResourceBinding(%s)", name)
+		err = s.scheduleClusterResourceBinding(crb)
+		metrics.BindingSchedule(string(FirstSchedule), metrics.SinceInSeconds(start), err)
+		return err
+	}
+	policyPlacement, policyPlacementStr, err := s.getClusterPlacement(crb)
+	if err != nil {
+		return err
 	}
-	resourceBinding, err := s.bindingLister.ResourceBindings(ns).Get(name)
-	if apierrors.IsNotFound(err) {
+	appliedPlacement := util.GetLabelValue(crb.Annotations, util.PolicyPlacementAnnotation)
+	if policyPlacementStr != appliedPlacement {
+		// policy placement changed, need reschedule
+		klog.Infof("Reschedule ClusterResourceBinding(%s) as placement changed", name)
+		err = s.scheduleClusterResourceBinding(crb)
+		metrics.BindingSchedule(string(ReconcileSchedule), metrics.SinceInSeconds(start), err)
+		return err
+	}
+	if policyPlacement.ReplicaScheduling != nil && util.IsBindingReplicasChanged(&crb.Spec, policyPlacement.ReplicaScheduling) {
+		// binding replicas changed, need reschedule
+		klog.Infof("Reschedule ClusterResourceBinding(%s) as replicas scaled down or scaled up", name)
+		err = s.scaleScheduleClusterResourceBinding(crb)
+		metrics.BindingSchedule(string(ScaleSchedule), metrics.SinceInSeconds(start), err)
+		return err
+	}
+	// TODO(dddddai): reschedule bindings on cluster change
+	if s.allClustersInReadyState(crb.Spec.Clusters) {
+		klog.Infof("Don't need to schedule ClusterResourceBinding(%s)", name)
 		return nil
 	}
-
-	return s.scheduleResourceBinding(resourceBinding)
+	if features.FeatureGate.Enabled(features.Failover) {
+		klog.Infof("Reschedule ClusterResourceBinding(%s) as cluster failure", name)
+		err = s.rescheduleClusterResourceBinding(crb)
+		metrics.BindingSchedule(string(FailoverSchedule), metrics.SinceInSeconds(start), err)
+		return err
+	}
+	return nil
 }
 
 func (s *Scheduler) scheduleResourceBinding(resourceBinding *workv1alpha2.ResourceBinding) (err error) {
+	klog.V(4).Infof("begin scheduling ResourceBinding %s/%s", resourceBinding.Namespace, resourceBinding.Name)
```

Comment:
I think this could be more concise. How do you think?
```
klog.V(4).InfoS("Begin scheduling resource binding", "resourceBinding", klog.KObj(resourceBinding))
```
It's ok. You could do your way. Just some tips.
By the way, the message of `klog` must begin with an uppercase letter.

Reply:
Thanks for your suggestion, done, PTAL :)